examples/example_unrolling_service/service_py/example_service.py

```diff
@@ -0,0 +1,162 @@
+"""An example CompilerGym service in python."""
+import logging
+import subprocess
+import sys
+from pathlib import Path
+from signal import Signals
+from typing import List, Optional, Tuple
+
+from compiler_gym.service import CompilationSession
+from compiler_gym.service.proto import (
+    Action,
+    ActionSpace,
+    Benchmark,
+    BenchmarkInitError,
+    Observation,
+    ObservationSpace,
+    ScalarLimit,
+    ScalarRange,
+    ScalarRangeList,
+)
+from compiler_gym.service.runtime import create_and_run_compiler_gym_service
+
+
+class UnrollingCompilationSession(CompilationSession):
+    """Represents an instance of an interactive compilation session."""
+
+    compiler_version: str = "1.0.0"
+
+    # The list of actions that are supported by this service. This example uses
+    # a static (unchanging) action space, but this could be extended to support
+    # a dynamic action space.
+    action_spaces = [
+        ActionSpace(
+            name="default",
+            action=[
+                "a",
+                "b",
+                "c",
+            ],
+        )
+    ]
+
+    # A list of observation spaces supported by this service. Each of these
+    # ObservationSpace protos describes an observation space.
+    observation_spaces = [
+        ObservationSpace(
+            name="ir",
+            string_size_range=ScalarRange(min=ScalarLimit(value=0)),
+            deterministic=True,
+            platform_dependent=False,
+            default_value=Observation(string_value=""),
+        ),
+        ObservationSpace(
+            name="features",
+            int64_range_list=ScalarRangeList(
+                range=[
+                    ScalarRange(
+                        min=ScalarLimit(value=-100), max=ScalarLimit(value=100)
+                    ),
+                    ScalarRange(
+                        min=ScalarLimit(value=-100), max=ScalarLimit(value=100)
+                    ),
+                    ScalarRange(
+                        min=ScalarLimit(value=-100), max=ScalarLimit(value=100)
+                    ),
+                ]
+            ),
+        ),
+        ObservationSpace(
+            name="runtime",
+            scalar_double_range=ScalarRange(min=ScalarLimit(value=0)),
+            deterministic=False,
+            platform_dependent=True,
+            default_value=Observation(
+                scalar_double=0,
+            ),
+        ),
+    ]
+
+    def __init__(
+        self, working_directory: Path, action_space: ActionSpace, benchmark: Benchmark
+    ):
+        super().__init__(working_directory, action_space, benchmark)
+        logging.info("Started a compilation session for %s", benchmark.uri)
+        self._benchmark = benchmark
+        self._update_observations()
+
+    def _update_observations(self):
+        self._observation = dict()
+        # FIXME: "llvm-dis" path is not found. Should we add its path, or is there another way to obtain the IR?
```

Comment:
In case you haven't seen, [this module](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/third_party/llvm/__init__.py#L85-L87) contains some useful functions for getting paths to LLVM tools, although we currently don't ship llvm-dis.